pkg/gofr/datasource/opentsdb/preprocess.go

```diff
@@ -0,0 +1,395 @@
+	Timestamp int64 `json:"timestamp"`
+
+	// The real type of Value only could be int, int64, float64, or string, and is required.
+	Value any `json:"value"`
+
+	// A map of tag name/tag value pairs. At least one pair must be supplied.
+	// Don't use too many tags, keep it to a fairly small number, usually up to 4 or 5 tags
+	// (By default, OpenTSDB supports a maximum of 8 tags, which can be modified by add
+	// configuration item 'tsd.storage.max_tags' in opentsdb.conf).
+	Tags map[string]string `json:"tags"`
+}
+
+// PutError holds the error message for each putting DataPoint instance. Only calling PUT() with "details"
+// query parameter, the response of the failed put data operation can contain an array PutError instance
+// to show the details for each failure.
+type PutError struct {
+	Data     DataPoint `json:"datapoint"`
+	ErrorMsg string    `json:"error"`
+}
+
+// PutResponse acts as the implementation of Response in the /api/put scene.
+// It holds the status code and the response values defined in
+// the [OpenTSDB Official Docs]: http://opentsdb.net/docs/build/html/api_http/put.html.
+type PutResponse struct {
+	Failed  int64      `json:"failed"`
+	Success int64      `json:"success"`
+	Errors  []PutError `json:"errors,omitempty"`
+}
+
+func (c *Client) getResponse(ctx context.Context, putEndpoint string, datapoints []DataPoint,
+	message *string) (*PutResponse, error) {
+	marshaled, err := json.Marshal(datapoints)
+	if err != nil {
+		*message = fmt.Sprintf("getPutBodyContents error: %s", err)
+		c.logger.Errorf(*message)
+	}
+
+	reqBodyCnt := string(marshaled)
+
+	putResp := PutResponse{}
+
+	if err = c.sendRequest(ctx, http.MethodPost, putEndpoint, reqBodyCnt, &putResp); err != nil {
+		*message = fmt.Sprintf("error processing put request at url %q: %s", putEndpoint, err)
+		return nil, err
+	}
+
+	return &putResp, nil
+}
+
+func parsePutErrorMsg(resp *PutResponse) error {
+	buf := bytes.Buffer{}
+	buf.WriteString(fmt.Sprintf("Failed to put %d datapoint(s) into opentsdb \n", resp.Failed))
+
+	if len(resp.Errors) > 0 {
+		for _, putError := range resp.Errors {
+			str, err := json.Marshal(putError)
+			if err != nil {
+				return err
+			}
+
+			buf.WriteString(fmt.Sprintf("\t%s\n", str))
+		}
+	}
+
+	return errors.New(buf.String())
+}
+
+func validateDataPoint(datas []DataPoint) error {
+	if len(datas) == 0 {
+		return errors.New("the given datapoint is empty")
+	}
+
+	for _, data := range datas {
+		if !isValidDataPoint(&data) {
+			return errors.New("the value of the given datapoint is invalid")
+		}
+	}
+
+	return nil
+}
+
+func isValidDataPoint(data *DataPoint) bool {
+	if data.Metric == "" || data.Timestamp == 0 || len(data.Tags) < 1 || data.Value == nil {
+		return false
+	}
+
+	switch data.Value.(type) {
+	case int64:
+		return true
+	case int:
+		return true
+	case float64:
+		return true
+	case float32:
+		return true
+	case string:
+		return true
+	default:
+		return false
+	}
```

Comment:
Maybe this


```suggestion
	switch data.Value.(type) {
	case int64:
	case int:
	case float64:
	case float32:
	case string:
	default:
		return false
	}
	return true
```

I don't think we can do that with a switch type

```suggestion
	switch data.Value.(type) {
	case int64, int, float64, float32, string:
	default:
		return false
	}
	return true
```